File: cmd/par/totientRange.go
// totientRange.go -- Sequential Euler Totient Function (Go Version)
// compile -- go build
// run --    totientRange lower_num upper_num

// Based on code from Phil Trinder from earlier work by: Greg Michaelson,
// Patrick Maier, Phil Trinder, Nathan Charles, Hans-Wolfgang Loidl and Colin
// Runciman

// This program calculates the sum of the totients between a lower and an upper
// limit

package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"time"
)

// Compute the Highest Common Factor, hcf of two numbers x and y
// hcf x 0 = x
// hcf x y = hcf y (mod x y)
func hcf(x, y int64) int64 {
	var t int64
	for y != 0 {
		t = x % y
		x = y
		y = t
	}
	return x
}

// relprime x y = hcf x y == 1
func relprime(x, y int64) bool {
	return hcf(x, y) == 1
}

func sumTotientPar(lower, upper int64) int64 {

	numWorkers := runtime.NumCPU()

	jobs := make(chan int64, upper-lower)
	results := make(chan int64, upper-lower)

	var wg sync.WaitGroup

	for w := 0; w < numWorkers; w++ {
		wg.Add(1)
		go eulerWorker(jobs, results, &wg)
	}

	go func() {
		for i := lower; i <= upper; i++ {
			jobs <- i
		}
		close(jobs)
	}()

	go func() {
		wg.Wait()
		close(results)
	}()

	var sum int64
	for r := range results {
		sum += r
	}

	return sum

}

func eulerWorker(jobs <-chan int64, results chan<- int64, wg *sync.WaitGroup) {
	defer wg.Done()
	var count, i int64

	for n := range jobs {
		count = 0
		for i = 1; i < n; i++ {
			if relprime(n, i) {
				count++
			}
		}
		results <- count
	}
}

func main() {
	var lower, upper int64
	var err error
	// Read and validate lower and upper arguments
	if len(os.Args) < 3 {
		panic(fmt.Sprintf("Usage: must provide lower and upper range limits as arguments"))
	}

	// Go if supports "If with a short statement"
	if lower, err = strconv.ParseInt(os.Args[1], 10, 64); err != nil {
		panic(fmt.Sprintf("Can't parse first argument"))
	}
	if upper, err = strconv.ParseInt(os.Args[2], 10, 64); err != nil {
		panic(fmt.Sprintf("Can't parse second argument"))
	}

	start := time.Now()
	totients := sumTotientPar(lower, upper)
	elapsed := time.Since(start)
	fmt.Println("Sum of Totients between", lower, "and", upper, "is", totients)
	fmt.Println("Elapsed time", elapsed)
}

File: c/src/par/totientRange.c
// TotientRange.c - Sequential Euler Totient Function (C Version)
// compile: gcc -Wall -O3 -o TotientRange TotientRange.c
// run:     ./TotientRange lower_num upper_num

// Based on code from Phil Trinder from earlier work by: Greg Michaelson,
// Patrick Maier, Phil Trinder, Nathan Charles, Hans-Wolfgang Loidl and Colin
// Runciman

// This program calculates the sum of the totients between a lower and an upper
// limit

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <time.h>
#include <omp.h>

// Compute the Highest Common Factor, hcf of two numbers x and y
// hcf x 0 = x
// hcf x y = hcf y (rem x y)
int64_t hcf(int64_t x, int64_t y) {
  int64_t t;
  while (y != 0) {
    t = x % y;
    x = y;
    y = t;
  }
  return x;
}

// relprime x y = (hcf x y == 1)
int relprime(int64_t x, int64_t y) {
  return hcf(x, y) == 1;
}

// euler n = length (filter (relprime n) [1 .. n-1])
int64_t euler(int64_t n) {
  int64_t length = 0;
  #pragma omp parallel reduction(+:length)
  {
    #pragma omp for schedule(dynamic, 100)
    for (int64_t i = 1; i < n; i++) {
      if (relprime(n, i)) {
        length++;
      }
    }
  }
  return length;
}


// sumTotient lower upper = sum (map euler [lower, lower+1 .. upper])
int64_t sumTotient(int64_t lower, int64_t upper) {
  int64_t sum = 0;
  #pragma omp parallel reduction(+:sum)
  {
    #pragma omp for schedule(dynamic, 1)
    for (int64_t i = lower; i <= upper; i++) {
      sum = sum + euler(i);
    }
  }
  return sum;
}


int main(int argc, char ** argv) {
  int64_t lower, upper;

  if (argc != 3) {
    printf("not 2 arguments\n");
    return 1;
  }

  sscanf(argv[1], "%" SCNd64, &lower);
  sscanf(argv[2], "%" SCNd64, &upper);

  struct timespec start, end;
  clock_gettime(CLOCK_MONOTONIC, &start);
  int64_t sum = sumTotient(lower, upper);
  clock_gettime(CLOCK_MONOTONIC, &end);

  double time_taken_ns = 1000000000L * (end.tv_sec - start.tv_sec) + end.tv_nsec - start.tv_nsec;

  printf("C: Sum of Totients  between [%ld..%ld] is %ld\n",
         lower, upper, sum);
  printf("Time Taken: %f s\n", time_taken_ns / 1000 / 1000 / 1000);

  return 0;
}
